R/modules.R

```diff
@@ -321,11 +321,124 @@ modules <- function(..., label = "root") {
 # printing methods ----
 
 #' @rdname teal_modules
+#' @param what character vector specifying which metadata to display. 
+#'   Possible values: "data_sources", "properties", "ui_args", "server_args", "transformers"
 #' @export
-format.teal_module <- function(x, indent = 0, ...) {
-  paste0(paste(rep(" ", indent), collapse = ""), "+ ", x$label, "\n", collapse = "")
+format.teal_module <- function(x, indent = 0, is_last = FALSE, parent_prefix = "", what = c("data_sources", "properties", "ui_args", "server_args", "transformers"), ...) {
+  empty_text <- "─ none"
+  branch <- if (is_last) "└─" else "├─"
+  current_prefix <- paste0(parent_prefix, branch, " ")
+  content_prefix <- paste0(parent_prefix, if (is_last) "   " else "│  ")
+
+  format_list <- function(lst, empty = empty_text) {
+    if (is.null(lst) || length(lst) == 0) {
+      empty
+    } else {
+      paste(names(lst), collapse = ", ")
+    }
+  }
+
+  bookmarkable <- isTRUE(attr(x, "teal_bookmarkable"))
+  reportable <- "reporter" %in% names(formals(x$server))
+
+  transformers <- if (length(x$transformers) > 0) {
+    paste(sapply(x$transformers, function(t) attr(t, "label")), collapse = ", ")
+  } else {
+    empty_text
+  }
+
+  output <- paste0(current_prefix, crayon::bold(x$label), "\n")
+  
+  if ("data_sources" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::blue("Data Sources:"), " ", paste(x$datanames, collapse = ", "), "\n"
+    )
+  }
+  if ("properties" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::yellow("Properties:"), "\n",
+      content_prefix, "│  ├─ Bookmarkable: ", bookmarkable, "\n",
+      content_prefix, "│  └─ Reportable: ", reportable, "\n"
+    )
+  }
+  if ("ui_args" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::green("UI Arguments:"), " ", format_list(x$ui_args), "\n"
+    )
+  }
+  if ("server_args" %in% what) {
+    output <- paste0(
+      output,
+      content_prefix, "├─ ", crayon::green("Server Arguments:"), " ", format_list(x$server_args), "\n"
```

Comment:
If we can make `UI Arguments` and `Server Arguments` to have the same amount of characters, then this output would be a  bit more readable:



<img width="268" alt="image" src="https://github.com/user-attachments/assets/529cd0a5-c089-40c7-8492-5a1138be7e7c">

Reply:
Or maybe we can adjust `:` so they are in the same column, the same way as it is in `print.teal_slice`

<img width="267" alt="image" src="https://github.com/user-attachments/assets/362d65e4-6fb1-46f4-93be-2fa8890264b2">

Reply:
Good point, I think it also holds true for the sub divisions of the Properties.

Reply:
Done, only thing bothering me is the wrap, without that it overflows terribly.
![image](https://github.com/user-attachments/assets/a784003b-e000-4942-8a15-5721104e762a)

Reply:
That's why I think we should control for the width and trim the width, so there is no overlap

Reply:
Or UI Arguments  / Server Arguments can be printed vertically instead of horizontally.

So instead of

```
  │  │  ├─ UI Arguments: plot_type, height, width, color_scheme
  │  │  ├─ Server Arguments: render_type, cache_plots
```

we get

```
  │  │  ├─ UI Arguments: 
                       - plot_type, 
                       - height, 
                       - width, 
                       - color_scheme
  │  │  ├─ Server Arguments: 
                       - render_type, 
                       - cache_plots
```

Reply:
Thanks! This change is implemented now. Great suggestion!